src/v/storage/disk_log_impl.cc

```diff
@@ -82,6 +82,8 @@ disk_log_impl::~disk_log_impl() {
 ss::future<> disk_log_impl::remove() {
     vassert(!_closed, "Invalid double closing of log - {}", *this);
     _closed = true;
+    // wait for compaction to finish
+    co_await _compaction_gate.close();
```

Comment:
I wonder if this is related to https://github.com/redpanda-data/redpanda/issues/3884 as well? (CC @VadimPlh )

Reply:
this caused `use-after-free` error

Reply:
I think it will fix #3884.